Source/WebKit/UIProcess/Automation/BidiScriptAgent.cpp

```diff
@@ -141,12 +155,311 @@ void BidiScriptAgent::evaluate(const String& expression, bool awaitPromise, Ref<
+
+    if (contextHandle.isEmpty()) {
+        // Fallback to frame-based ID if we can't get context handle
+        newRealmId = makeString("realm-frame-"_s, String::number(frameInfo.frameID.toUInt64()));
+    } else {
+        // Use the contextHandle directly - it's already unique for both main frames and iframes
+        // For the first load of a context, use just the context handle
+        // For subsequent navigations/reloads, append a counter to make it unique
+        auto counterIt = m_frameRealmCounters.find(frameInfo.frameID);
+        if (counterIt == m_frameRealmCounters.end()) {
+            // First realm for this frame - no counter suffix
+            newRealmId = makeString("realm-"_s, contextHandle);
+            // Start counter at 1 so the NEXT navigation will use "-1" suffix
+            m_frameRealmCounters.set(frameInfo.frameID, 1);
+        } else {
+            // Subsequent realm (reload/navigation) - use and increment counter
+            uint64_t counter = counterIt->value;
+            newRealmId = makeString("realm-"_s, contextHandle, "-"_s, String::number(counter));
+            counterIt->value = counter + 1;
+        }
+    }
+
+    // Update the cache with the new state signature and realm ID
+    m_frameRealmCache.set(frameInfo.frameID, std::make_pair(stateSignature, newRealmId));
+
+    return newRealmId;
+}
+
+String BidiScriptAgent::generateRealmIdForBrowsingContext(const String& browsingContext)
+{
+    // For evaluate/callFunction, we need to generate consistent realm IDs based on the browsing context.
+    // This simplified version works for main window contexts (page handles).
+    // For now, we just use the browsing context handle as the realm ID base.
+    // This will match what getRealms() generates for main frames since contextHandleForFrame returns the page handle.
+
+    // The realm ID should match the format used by generateRealmIdForFrame()
+    return makeString("realm-"_s, browsingContext);
+}
+
+String BidiScriptAgent::originStringFromSecurityOriginData(const WebCore::SecurityOriginData& originData)
+{
+    if (originData.isOpaque())
+        return "null"_s;
+
+    return originData.toString();
+}
+
+void BidiScriptAgent::processRealmsForPagesAsync(Vector<Ref<WebPageProxy>>&& pagesToProcess, std::optional<Inspector::Protocol::BidiScript::RealmType>&& optionalType, std::optional<String>&& contextHandleFilter, Vector<RefPtr<Inspector::Protocol::BidiScript::RealmInfo>>&& accumulated, Inspector::CommandCallback<Ref<JSON::ArrayOf<Inspector::Protocol::BidiScript::RealmInfo>>>&& callback)
+{
+    if (pagesToProcess.isEmpty()) {
+        // Assemble final array with window realms only
+        auto realmsArray = JSON::ArrayOf<Inspector::Protocol::BidiScript::RealmInfo>::create();
+        for (auto& realmInfo : accumulated) {
+            if (!realmInfo)
+                continue;
+            if (optionalType && *optionalType != Inspector::Protocol::BidiScript::RealmType::Window)
```

Comment:
`std::optional<T>::operator!=` does this for you actually
```suggestion
            if (optionalType != Inspector::Protocol::BidiScript::RealmType::Window)
```

Reply:
I was following that part of the BiDi spec [here](https://w3c.github.io/webdriver-bidi/#command-script-getRealms) :

> If command parameters contains type and realm info["type"] is not equal to command parameters["type"] then continue.

In the case we have no type given in input (Type = nullopt), if we use the operator  `std::optional<T>::operator!=`  the line 366 will be true ( nullopt != Window return true) and no realm will be output.
The current implementation follow the path such that if no type is given as input, we should output all the realms (no filtering). I think it follows the BiDi two parts condition to the letter.

Reply:
ah okay, I misunderstood that intent.